bundle/src/main/java/com/adobe/acs/commons/wcm/vanity/impl/VanityURLServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+
+	@Reference
+	QueryBuilder queryBuilder;
+
+	public boolean dispatch(SlingHttpServletRequest request, SlingHttpServletResponse response) throws ServletException, IOException, RepositoryException {
+		if (request.getAttribute(VANITY_DISPATCH_CHECK_ATTR) != null) {
+			log.trace("Processing a previously vanity dispatched request. Skipping...");
+			return false;
+		}
+
+		request.setAttribute(VANITY_DISPATCH_CHECK_ATTR, true);
+
+		final String requestURI = request.getRequestURI();
+		final RequestPathInfo requestPathInfo = new PathInfo(request.getResourceResolver(), requestURI);
+
+		// Manually strip off any selectors or extensions from the URL
+		final String resourcePath = StringUtils.substringBefore(requestPathInfo.getResourcePath(), ".");
+		// Map the incoming URL to remove any prefix
+		final String candidateVanity = request.getResourceResolver().map(resourcePath);
+		final String pathScope = StringUtils.removeEnd(resourcePath, candidateVanity);
+
+		log.debug("Candidate vanity URL to check and dispatch: [ {} ]", candidateVanity);
+
+		// Check if...
+		// 1) the candidateVanity and the requestURI are the same; If they are it means the request has already
+		// gone through resource resolution and failed so there is no sense in sending it through again.
+		// 2) the candidate is in at least 1 sling:vanityPath under /content
+		if (!StringUtils.equals(candidateVanity, requestURI) && isVanityPath(pathScope, candidateVanity, request)) {
+			log.debug("Forwarding request to vanity resource [ {} ]", candidateVanity);
+
+			final RequestDispatcher requestDispatcher = request.getRequestDispatcher(candidateVanity);
+			requestDispatcher.forward(new ExtensionlessRequestWrapper(request), response);
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks if the provided vanity path is sling:vanityPath under the mapped path prefix
+	 *
+	 * @param pathScope The content path to scope the vanity search too.
+	 * @param vanityPath Vanity path that needs to be validated.
+	 * @param request SlingHttpServletRequest object used for performing query/lookup
+	 * @return return true if the vanityPath is a registered sling:vanityPath under /content
+	 */
+	protected boolean isVanityPath(String pathScope, String vanityPath, SlingHttpServletRequest request) throws RepositoryException {
+		final long start = System.currentTimeMillis();
+
+		final Map<String, String> params = new HashMap<>();
+
+		// Limit to <pathScope> to get 1/2 to multi-tenant support
+		params.put("path", StringUtils.defaultIfEmpty(pathScope, DEFAULT_PATH_SCOPE));
+		params.put("property", NameConstants.PN_SLING_VANITY_PATH);
+		params.put("property.value", vanityPath);
+		params.put("p.limit", "1");
+		params.put("p.guessTotal", "true");
+
+		log.debug("Searching for vanity path [ {} ] under [ {} ]", vanityPath, pathScope);
+
+		final Query query = queryBuilder.createQuery(PredicateGroup.create(params), request.getResourceResolver().adaptTo(Session.class));
```

Comment:
If I understood correctly, this query is executed on every request. Is my understanding correct? In that case this would have a huge performance impact even if this query is backed by an optimized Oak index.